src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.Automata.cs

```diff
@@ -266,17 +385,21 @@ private int GetCoreStateId(int nfaStateId)
         /// <summary>Gets or creates a new DFA transition.</summary>
         /// <remarks>This function locks the matcher for safe concurrent use of the <see cref="_builder"/></remarks>
         private bool TryCreateNewTransition(
-            MatchingState<TSet> sourceState, int mintermId, int offset, bool checkThreshold, [NotNullWhen(true)] out MatchingState<TSet>? nextState)
+            MatchingState<TSet> sourceState, int mintermId, int offset, bool checkThreshold, [NotNullWhen(true)] out MatchingState<TSet>? nextState,
+            long timeoutOccursAt = 0)
         {
             Debug.Assert(offset < _dfaDelta.Length);
-
             lock (this)
             {
                 // check if meanwhile delta[offset] has become defined possibly by another thread
                 MatchingState<TSet>? targetState = _stateArray[_dfaDelta[offset]];
                 if (targetState is null)
                 {
-                    if (checkThreshold && _stateCache.Count >= SymbolicRegexThresholds.NfaThreshold)
+                    if (// check if there is an active timer
+                        (timeoutOccursAt != 0 && Environment.TickCount64 > timeoutOccursAt) ||
```

Comment:
Since this part hasn't been explained yet:
The timeout was passed down here because it's the most time-consuming part of the match loop. Doing some tests with massive patterns and a 1s timeout i found that the real timeout was triggered somewhere around 20s, because the timeout check was for a fixed number of characters traversed, where each character took a long time creating a new state here. This ensures that the timeout exception throws much closer to the specified time.

Reply:
worth comment?

Reply:
Yes this is in addition to the batch of characters. This branch stops being called once the engine is fully built